cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,491 @@
+        """
+            A shorthand to (efficiently) compute all (optimal) solutions, map them to CPMpy and optionally display the solutions.
+
+            If the problem is an optimization problem, returns only optimal solutions.
+
+           Arguments:
+                - display: either a list of CPMpy expressions, OR a callback function, called with the variables after value-mapping
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - call_from_model: whether the method is called from a CPMpy Model instance or not
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+
+        import docplex.cp as docp
+        #cpo_solver = self.cpo_model.start_search(TimeLimit=time_limit, SolutionLimit=solution_limit, LogVerbosity='Quiet', **kwargs)
+        solution_count = 0
+        if solution_limit is None:
+            solution_limit = 99999
+        while solution_count < solution_limit:
+            cpo_result = self.solve(time_limit=time_limit, **kwargs)
+            if not cpo_result:
+                break
+            solution_count += 1
+            '''if cpo_result.is_solution_optimal():
+                print('optimal')
+            if not cpo_result.is_new_solution():
+                break
+            solution_count += 1'''
+            if self.has_objective():
+                # only find all optimal solutions
+                self.cpo_model.add(self.cpo_model.get_objective_expression().children[0] == self.objective_value_)
+            solvars = []
+            vals = []
+            for cpm_var in self.user_vars:
+                sol_var = self.solver_var(cpm_var)
+                cpm_value = cpm_var._value
+                if isinstance(cpm_var, _BoolVarImpl):
+                    # because cp optimizer does not have boolean variables we use an integer var x with domain 0, 1
+                    # and then replace a boolvar by x == 1
+                    sol_var = sol_var.children[0]
+                    cpm_value = int(cpm_value)
+                solvars.append(sol_var)
+                vals.append(cpm_value)
+            self.cpo_model.add(docp.modeler.forbidden_assignments(solvars, [vals]))
+            if display is not None:
+                if isinstance(display, Expression):
+                    print(argval(display))
+                elif isinstance(display, list):
+                    print(argvals(display))
+                else:
+                    display()  # callback
+        return solution_count
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        import docplex.cp as docp
```

Comment:
this might hurt performance.... we are going to call `solver_var()` many many times, and if each requires an import check...

Reply:
gurobi also has an import here